TSOClient/FSO.Server.Api.Core/Controllers/UserOAuthController.cs

```diff
@@ -0,0 +1,128 @@
+using Microsoft.AspNetCore.Mvc;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Web;
+using System.Web.Http;
+
+namespace FSO.Server.Api.Core.Controllers
+{
+    [Route("userapi/oauth/token")]
+    [ApiController]
+    public class UserOAuthController : ControllerBase
+    {
+        [HttpPost]
+        public IActionResult CreateToken([FromForm] UserOAuthRequest userAuthRequest)
+        {
+            if (userAuthRequest == null) BadRequest();
+            var api = Api.INSTANCE;
+            using (var da = api.DAFactory.Get())
+            {
+                var user = da.Users.GetByUsername(userAuthRequest.username);
+                if (user == null || user.is_banned) return ApiResponse.Json(System.Net.HttpStatusCode.Unauthorized, new UserOAuthError("unauthorized_client", "user_credentials_invalid"));
+                var ip = ApiUtils.GetIP(Request);
+                var hashSettings = da.Users.GetAuthenticationSettings(user.user_id);
+                var isPasswordCorrect = PasswordHasher.Verify(userAuthRequest.password, new PasswordHash
+                {
+                    data = hashSettings.data,
+                    scheme = hashSettings.scheme_class
+                });
+                //check if account is locked due to failed attempts
+                var accLock = da.Users.GetRemainingAuth(user.user_id, ip);
+                if (accLock != null && (accLock.active || accLock.count >= AuthLoginController.LockAttempts) && accLock.expire_time > Epoch.Now)
+                {
+                    return ApiResponse.Json(System.Net.HttpStatusCode.OK, new UserOAuthError("unauthorized_client", "account_locked"));
+                }
+                //if the password is incorrect and check if user failed muli times and set a time out till next try.
+                if (!isPasswordCorrect)
+                {
+                    var durations = AuthLoginController.LockDuration;
+                    var failDelay = 60 * durations[Math.Min(durations.Length - 1, da.Users.FailedConsecutive(user.user_id, ip))];
+                    if (accLock == null)
+                    {
+                        da.Users.NewFailedAuth(user.user_id, ip, (uint)failDelay);
+                    }
+                    else
+                    {
+                        var remaining = da.Users.FailedAuth(accLock.attempt_id, (uint)failDelay, AuthLoginController.LockAttempts);
+                    }
+                    return ApiResponse.Json(System.Net.HttpStatusCode.OK, new UserOAuthError("unauthorized_client", "user_credentials_invalid"));
+                }
+
+                //user passed the password check, and now creates the claim/token
+                da.Users.SuccessfulAuth(user.user_id, ip);
+                var claims = new List<string>();
+
+                //set the permission level in the claim
+                switch (userAuthRequest.permission_level)
```

Comment:
What will be the extent of these permissions? They seem to be granted immediately to any user authenticating, and obviously right now the only one that works is reading avatars for the logged user.